Flush queue and "SafeBufferedAppender"

GregAscolab opened the pull request:
Two points that my usage need :   
* Ability to flush `BufferedAppender` periodically (instead of only when new message to append)  
* Fix a potential lost of item in `BufferedAppender` : don't remove the item from the ring buffer if it failed to be "append" !  

So I add a `autoFlushPeriod` in `LogQueue` class, and then send `nullptr` messages to appenders to trigger the flush loop.  

And I create a "copy" of `BufferedAppender`, to change the behavior and avoid loosing messages.
This `SafeBufferedAppender` take care to not remove item if it has not really been sent.
Also include a `maxLoopItems`, to avoid infinite loop, that could take too much time of the Queue task and raised Watchdog interrupt! (When `useQueue`)

It's my first contribution to open source project, don't hesitated to comment and change necessary code if needed.

Commit:
Fix comment

Commit:
Add a flush period of BufferedAppender, when useQueue.

The autoFlushPeriod parameter allow the "Queue" task to be run periodically, even if there is no new item to append to appenders.
It add a "timeout" period to the blocking "xRingbufferReceive", and in case of no new message, it will loop on all appenders, appending "nullptr", to make them flushing their buffers (in case of buffered appenders ;-) )

Commit:
Add missing xSemaphoreTake + comment

Seems missing taking the semaphore before returning item to ring buffer (like other place in code...)
+
Add comment on potential lost of items...

Commit:
Create new SafeBufferedAppender

Same principle as BufferedAppender, but fix a potential issue of loosing item !
Here we alway add message to buffer, and then try to flush the buffer.
If item is not sent, we keep reference in memory and try next time.

Could be used in addition with useQueue "autoFlushPeriod" for more convenience.

Commit:
Add comments

dyarkovoy commented:
Thank you for the PR, good catch with the missing xSemaphoreTake, no objections on the flush period for BufferedApender, and comment fixes. 
Regarding  SafeBufferedAppender, do you think we could add an option to the existing BufferedAppender to enable/disable your proposed functionality? I think it would be better than duplicating the functionality of SafeBufferedAppender in a separate class, what do you think?

GregAscolab commented:
Hi,
I agree concerning avoiding duplicating "BufferedAppender". It was my first idea... but I was not sure for retro compatibility.
But I think it could be done by "inserting" my code inside the "append" method of the BufferedAppender, tested by a simple "if/else".
Any idea for the option name ? `safe` or `memorizedItem` ?

GregAscolab commented:
What about `_max_loop_item_sent` ?
It's also need more tests concerning "autoRelease" option... (that I removed in SafeBufferedAppender)

dyarkovoy commented:
Let me understand better what you are trying to achieve with this new functionality for BufferedAppender. You say that it prevents loss of messages, but this is the idea behind BufferedAppender - to keep as many messages as the size of the buffer permits, until the appender is ready, overwriting the older messages as the buffer gets full. Are you saying that the existing BufferedAppender  doesn't keep the messages?

GregAscolab commented:
Exactly... there is a very small mistake/case where the BufferedAppender can loose message (I get the case ;-) ) :  

Just here.  
Between the `while` test, and the call to `_appender.append(item)`, something can happen...
So if `_appender.append(item)` return false, you "return the item" from ring buffer by calling `vRingbufferReturnItem(_handle, item);`, that will remove the item for the next `xRingbufferReceive` call !!!

But furtheremore, after this while loop, you  try to send the new message... and in that case it could possibly really be sent by the append() call (return true), so you don't add the message to the ring buffer...
So you mix the FIFO order !

So finaly : The new message will be sent before older, and you loose one message ;-)

```
while (_appender.append(nullptr))
            {
                xSemaphoreTake(_lock, portMAX_DELAY);
                item = (LogMessage *)xRingbufferReceive(_handle, &size, 0);
                xSemaphoreGive(_lock);
                if (!item)
                    break;
                ok &= _appender.append(item); // BE CAREFUL : If for any reason, append() is not ok, the item is lost... (even if the while loop test is ok, it can change until here...)
                xSemaphoreTake(_lock, portMAX_DELAY);
                vRingbufferReturnItem(_handle, item);
                xSemaphoreGive(_lock);
            }
            if (!_appender.append(message))
             ...
```

GregAscolab commented:
That's why I propose to :  
1. Always add message to ring buffer : make place if necessary. And of course overwrite the oldest message (the last one, that has not yet be sent) if we need space.
2. Try to flush ring buffer, keeping in memory the last item not sent.

Commit:
Fix the BufferedAppender behavior + add maxLoopItems option.

Finally replace the BufferedAppender append() method with new code to fix wrong behavior...
* Possible loose of message
* Possible mix in FIFO message order

New maxLoopItems option avoid sending too much item per append() call...
In case of useQueue, the Queue task could take too much time, and never trigger the watchdog !

GregAscolab commented:
So finally, if the behavior is ok for you, I fix the "append()" method with the new one.
(And remove the SafeBufferedAppender class)
Including the autoRelease and maxLoopItems option.

Have a look.

Commit:
Fix comments

dyarkovoy commented:
Okay, I see. I guess my original assumption was that if the appender succeeds once, it will always succeed. Which is not necessarily true. Let's try your solution.
